homeassistant/components/overkiz/config_flow.py

```diff
@@ -1,78 +1,203 @@
+        self._user = None
+        self._server = DEFAULT_SERVER
+        self._host = "gateway-xxxx-xxxx-xxxx.local:8443"
 
-    async def async_validate_input(self, user_input: dict[str, Any]) -> None:
+    async def async_validate_input(self, user_input: dict[str, Any]) -> dict[str, Any]:
         """Validate user credentials."""
         username = user_input[CONF_USERNAME]
         password = user_input[CONF_PASSWORD]
-        server = SUPPORTED_SERVERS[user_input[CONF_HUB]]
-        session = async_create_clientsession(self.hass)
-
-        client = OverkizClient(
-            username=username, password=password, server=server, session=session
-        )
 
-        await client.login(register_event_listener=False)
-
-        # Set first gateway id as unique id
+        user_input[CONF_API_TYPE] = self._api_type
+
+        # Local API
+        if self._api_type == APIType.LOCAL:
+            # Create session on Somfy cloud server to generate an access token for local API
+            session = async_create_clientsession(self.hass)
+            server = SUPPORTED_SERVERS[self._server]
+            client = self._create_cloud_client(
+                username=username, password=password, server=server
+            )
+
+            await client.login(register_event_listener=False)
+            gateways = await client.get_gateways()
+
+            gateway_id = ""
+            for gateway in gateways:
+                # Overkiz can return multiple gateways, but we only can generate a token
+                # for the main gateway.
+                if is_overkiz_gateway(gateway.id):
+                    gateway_id = gateway.id
+
+            developer_mode = await client.get_setup_option(
+                f"developerMode-{gateway_id}"
+            )
+
+            if developer_mode is None:
+                raise DeveloperModeDisabled
+
+            token = await client.generate_local_token(gateway_id)
+            await client.activate_local_token(
+                gateway_id=gateway_id, token=token, label="Home Assistant/local"
+            )
+
+            # Verify SSL blocked by https://github.com/Somfy-Developer/Somfy-TaHoma-Developer-Mode/issues/5
+            # Somfy (self-signed) SSL cert uses the wrong common name
+            session = async_create_clientsession(self.hass, verify_ssl=False)
+
+            # Local API
+            local_client = OverkizClient(
+                username="",
+                password="",
+                token=token,
+                session=session,
+                server=generate_local_server(host=user_input[CONF_HOST]),
+            )
+
+            await local_client.login()
+
+            user_input[CONF_TOKEN] = token
+        else:
+            server = SUPPORTED_SERVERS[user_input[CONF_SERVER]]
+            client = self._create_cloud_client(
+                username=username, password=password, server=server
+            )
+
+            await client.login(register_event_listener=False)
```

Comment:
This might be a bit easier to maintain if you split this into a local and remote function